### PR TITLE
[Strings] Avoid mishandling unicode in interpreter

### PR DESCRIPTION
**Pull request by tlively**
Our interpreter implementations of `stringview_wtf16.length`,
`stringview_wtf16.get_codeunit`, and `string.encode_wtf16_array` are not
unicode-aware, so they were previously incorrect in the face of multi-byte code
units. As a fix, bail out of the interpretation if there is a non-ascii code
point that would make our naive implementation incorrect.

### CODE REVIEW

**Comment by tlively** [comment]
* **#6405** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6405?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tlively and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by tlively** [comment]
cc @rluble

File: test/lit/passes/precompute-strings.wast [hunk 1]
```diff
@@ -26,11 +29,104 @@
...
+    ;; $_£_€_𐍈
+    (string.const "$_\C2\A3_\E2\82\AC_\F0\90\8D\88")
+   )
+   (i32.const 2)
+  )
+ )
+
+ ;; CHECK:      (func $encode (type $0) (result i32)
+ ;; CHECK-NEXT:  (i32.const 2)
+ ;; CHECK-NEXT: )
```
Comment (tlively) [inline test/lit/passes/precompute-strings.wast:102]: @kripken, is this optimization safe? Does Precompute do this even if the modified array would escape?

File: src/wasm-interpreter.h [hunk 2]
```diff
@@ -1917,6 +1917,15 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
...
       trap("null ref");
     }
+
+    // This is only correct if all the bytes stored in `values` correspond to
+    // single unicode code points. See `visitStringWTF16Get` for details.
+    for (Index i = 0; i < data->values.size(); ++i) {
+      if (uint32_t(data->values[i].geti32()) > 127) {
+        return Flow(NONCONSTANT_FLOW);
+      }
+    }
```
Comment (kripken) [inline src/wasm-interpreter.h:1927]: Perhaps it makes sense to add a helper for this? Could be a templated function in `src/support/string.h`, or maybe a helper in this file as part of the interpreter would be better, I'm not sure.

**Review by kripken — COMMENTED** [review]
lgtm % suggestion

File: src/wasm-interpreter.h (hunk 2 above)
Comment (tlively) [reply src/wasm-interpreter.h:1935]: Since we're operating on lists of literals, I've added a helper to the interpreter.

**Review by kripken — APPROVED** [review]

**Pull request merged**